Feature/affine alignment 2

rrahn opened the pull request:
Depends on:
* #668 

Supersedes #636 

Implements the first version of an affine alignment kernel that computes the score.
It refines the alignment range and the two-way executor to be move only as they are dealing with a resource. Some small fixes of the single input pass view to allow default construction of it.
Refines the algorithm configuration to check pre conditions and throw if a configuration setting is yet not supported.

joergi-w commented on include/seqan3/alignment/pairwise/policy/affine_gap_policy.hpp:
```diff
@@ -0,0 +1,101 @@
+    /*!\brief Creates the cache used for affine gap computation.
+     * \tparam    gap_scheme_t The type of the gap scheme.
+     * \param[in] scheme       The configured gap sheme.
+     * \returns The created cache.
```
Please document the meaning of the tuple elements. For instance, you redefine gap_open here as open_score + gap_score, which is different from the conventions and should therefore be clearly described. 
I suggest to assign only gap_open here and in the above function (compute_cell) do the sum.

rrahn replied:
Well, no the convention is actually that opening a gap is gap cost plus additional gap open costs. In SeqAn2 this was different. So I do the add in the cache generation in order to save the instruction in the hot kernel above.

rrahn replied:
And yes as described before, I will make it a spearate class which is then documented and easier to follow.

h-2 replied:
maybe also not call it `setup`? I would assume that any member function called `setup` somehow changes the object, but this doesn't.

h-2 replied:
yeah, make_cache is better, although I still feel that it is something that should be local to the object and created automatically, but that's up to you.

rrahn replied:
I'll revisit this when I do more benchmarks. This is all detail code anyway so I can refactor it later without breaking API

joergi-w commented on include/seqan3/alignment/pairwise/align_pairwise.hpp:
```diff
@@ -71,19 +71,17 @@ namespace seqan3
     requires detail::is_type_specialisation_of_v<remove_cvref_t<alignment_config_t>, configuration> &&
              tuple_like_concept<value_type_t<std::ranges::iterator_t<std::remove_reference_t<sequence_t>>>>
-constexpr auto align_pairwise(sequence_t && seq, alignment_config_t && config)
+constexpr auto align_pairwise(sequence_t && seq, alignment_config_t const & config)
```
documentation missing (whole file)

rrahn replied:
Yes, but this comes to a later point.

joergi-w commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,170 @@
+            get<1>(res) = get<0>(*(std::ranges::end(last_col) - 1));
+            return res;
+        }
+        return res;
```
wrap in else {} statement

rrahn replied:
I think this is all valid according to our style guides

joergi-w commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,170 @@
+     * \param[in,out]  cache   The cache holding hot variables.
+     */
+    template <typename cache_t>
+    void initialize(cache_t & cache)
```
`initialize` sounds like a constructor of alignment_algorithm. 
Consider a more descriptive name, e.g. `fill_1st_column`.

rrahn replied:
Mhmm, I am not against it per sé, but don't really see any improvements. The text books call this step initialization and the function is documented and no user will ever touch it.

rrahn replied:
I will though make it private instead of protected.

joergi-w commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,170 @@
+     * \tparam        cache_t        The type of the cache.
+     * \param[in]     first_batch    The first sequence.
+     * \param[in]     second_batch   The second sequence.
+     * \param[in,out] cache          The cache holding hot variables.
```
Please describe the cache better: which values are the hot ones?
(maybe it is enough to link to the actual cache class where the details are given)

rrahn replied:
At the moment the cache is just internally defined by the gap policy but I have a tracking card to make this a separate class.

joergi-w commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -0,0 +1,189 @@
+    //!\}
+
+public:
+    //!\brief Tests whether the value type of `range_type` is a tuple with exactly 2 members.
```
As said before, this could be an std::pair. In that case this test would be omitted.

rrahn replied:
We cannot assume anything about the type here. So I cannot require that sequences always have to be passed as std::pair. For example using the ranges::view::zip (which is very likely in this context) will return a tuple not a pair.

joergi-w commented on include/seqan3/alignment/pairwise/select_result_type.hpp:
```diff
@@ -0,0 +1,80 @@
+                return align_result<type_list<uint32_t,
+                                              score_type,
+                                              alignment_coordinate>>{};
+            else if constexpr (configuration_t::template exists<align_cfg::result<with_begin_position_type>>())
```
use brackets for multi-line if statements

rrahn replied:
IIRC this follows our convention of single expression if-clauses don't necessarily require `{}`  🤔

rrahn replied:
I changed it

rrahn commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -0,0 +1,189 @@
+        static_assert(alignment_contract_t::expects_tuple_like_value_type(),
+                      "Alignment configuration error: "
+                      "The value type of the sequence ranges must fulfill the tupke_like_concept and must contain "
+                      "exactly 2 elements.");
```
As I described above. This assumption would be to invasive.

rrahn commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -0,0 +1,189 @@
+    template <typename kernel_t, typename config_t>
+    static constexpr auto configure_edit_distance(config_t const & cfg)
+    {
+        return kernel_t{edit_distance_wrapper<remove_cvref_t<config_t>>{cfg}};
```
I need to refactor it to make it work with dynamically set end-gap specifiers. But this comes in another PR.

rrahn commented on include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp:
```diff
@@ -604,6 +605,77 @@ pairwise_alignment_edit_distance_unbanded(database_t && database, query_t && que
+                                                       std::forward<second_batch_t>(second_batch),
+                                                       *cfg_ptr};
+        result_t res{};
+        return algo(res);
```
No RVO will kick in.

rrahn commented:
@h-2 can you look at this PR? I addressed all issues from @joergi-w

h-2 commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,168 @@
+     * The configuration is not passed to the function-call-operator of this function object, in order to avoid
+     * incompatible configurations between the passed configuration and the one used during configuration of this
+     * class. Further will the function object be stored in a std::function which requires copyable objects and
+     * in parallel executions the function object must be copied as well.
```
I don't really understand this. Why do you pass a config to this object at all, and not just to `operator()`?

h-2 commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,168 @@
+        {
+            this->init_column_cell(cell, cache);
+        });
+    }
```
a lot of `this->` here, why?

h-2 replied:
:thinking:

h-2 commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,168 @@
+private:
+
+    //!\brief The alignment configuration stored on the heap.
+    std::shared_ptr<config_t> cfg_ptr{};
```
why is the config a member of the type, but the cache a local variable? Are any of these things called in parallel?

h-2 replied:
It's just that then the burder is on the user to take care of the cache, right?

rrahn replied:
No? The user never sees nor touches any of this. This is all implementation specific stuff for library developers only.

h-2 commented on include/seqan3/alignment/pairwise/execution/execution_handler_sequential.hpp:
```diff
@@ -52,27 +52,18 @@ namespace seqan3::detail
-    execution_handler_sequential & operator=(execution_handler_sequential const &) = default;
-    execution_handler_sequential & operator=(execution_handler_sequential &&)      = default;
-    ~execution_handler_sequential()                                                = default;
-    //!}
```
why are these removed? If this class is supposed to be an aggregate, please make it a struct

h-2 commented on include/seqan3/alignment/pairwise/execution/execution_handler_sequential.hpp:
```diff
@@ -52,27 +52,18 @@ namespace seqan3::detail
+    void execute(std::function<func_result_t(first_batch_t const &, second_batch_t const &)> func,
+                  first_batch_t const & first_batch,
+                  second_batch_t const & second_batch,
+                  std::function<void(decltype(func(first_batch, second_batch)))> delegate)
```
still open

h-2 commented on include/seqan3/alignment/pairwise/policy/affine_gap_init_policy.hpp:
```diff
@@ -0,0 +1,103 @@
+    template <typename cell_t, typename cache_t>
+    constexpr auto init_origin_cell(cell_t & active_cell, cache_t & cache) const noexcept
+    {
+        using std::get;
```
what for?

h-2 replied:
ADL means only looking in associated namespaces (and friends and some other things). Writing `using ...` is not ADL, it means "consider the function from this namespace **in addition to ADL**". 
But you aren't call get anywhere anyway. And the structured bindings are explicitly required to not look at the current namespace or any `using ...` you might have done, they are only allowed to consider member get and ADL get.
If it doesn't compile without `using std::get;` than I am pretty sure it is a bug if it does with it.

h-2 commented on include/seqan3/alignment/pairwise/policy/affine_gap_policy.hpp:
```diff
@@ -0,0 +1,101 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI für molekulare Genetik
```
2019 everywhere

h-2 commented on include/seqan3/alignment/pairwise/policy/affine_gap_init_policy.hpp:
```diff
@@ -0,0 +1,103 @@
+        vt_score = gap_open + gap_extend;
+    }
+
+    /*!\brief Initializes the cells in the first column of the dynamic programming matrix.
```
in British English it's initiali**s**e

(often in your code!)

h-2 commented on include/seqan3/alignment/pairwise/policy/affine_gap_policy.hpp:
```diff
@@ -0,0 +1,101 @@
+        auto & [prev_cell, gap_open, gap_extend] = cache;
+        auto & [prev_score, vt_score] = prev_cell;
+
+        // can we use GCC's builtin saturation arithmetics?
```
add a TODO maybe with a link what this is about?

h-2 commented on include/seqan3/alignment/pairwise/policy/unbanded_dp_matrix_policy.hpp:
```diff
@@ -0,0 +1,84 @@
+    void allocate_score_matrix(first_batch_t && first_batch, second_batch_t && second_batch)
+    {
+        dimension_first_batch = std::ranges::size(std::forward<first_batch_t>(first_batch));
+        dimension_second_batch = std::ranges::size(std::forward<second_batch_t>(second_batch));
```
Why do you forward to size?

h-2 replied:
You wouldn't want the size function to take ownership and destry your ranges :wink

h-2 commented on include/seqan3/alignment/pairwise/policy/unbanded_dp_matrix_policy.hpp:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    //!\brief Returns the active column that is computed next.
+    auto active_column()
```
is "active == current" or is "active == next" :thinking:

h-2 commented on include/seqan3/alignment/pairwise/select_result_type.hpp:
```diff
@@ -0,0 +1,83 @@
+                                               std::tuple<std::vector<first_seq_value_type>,
+                                                          std::vector<second_seq_value_type>>>{};
+            else
+                return align_result_value_type<uint32_t, score_type>{};
```
braces around individual blocks because they are multiline

h-2 commented on include/seqan3/alignment/pairwise/select_result_type.hpp:
```diff
@@ -0,0 +1,83 @@
+//!\cond
+    requires is_type_specialisation_of_v<remove_cvref_t<configuration_t>, configuration>
+//!\endcond
+struct select_result_type
```
not so important but the name sounds very general...

h-2 commented on test/unit/alignment/pairwise/align_pairwise_test.cpp:
```diff
@@ -118,11 +123,12 @@ TEST(align_pairwise, multiple_rng_lvalue)
+    //     auto && [gap1, gap2] = res.trace();
+    //     EXPECT_EQ(std::string{gap1 | view::to_char}, "ACGTGATG--");
+    //     EXPECT_EQ(std::string{gap2 | view::to_char}, "A-GTGATACT");
+    // }
```
all actual tests commented out?

rrahn replied:
Yes they are going to be refined anyway with more code that I add. This PR is already old, and there are newer once, that resolve already parts of it.

h-2 commented on test/unit/alignment/pairwise/alignment_selector_test.cpp:
```diff
@@ -87,25 +90,38 @@ TEST(alignment_selector, determine_result_type)
+        // seq_t s{};
+        // EXPECT_TRUE((std::is_same_v<res_t, typename _t::result_type>));
+        // EXPECT_TRUE((std::is_same_v<decltype(_t{cfg}.select(s)),
+        //                             std::function<res_t(res_t &)>>));
```
...?

h-2 commented on test/unit/alignment/pairwise/global_affine_unbanded_test.cpp:
```diff
@@ -0,0 +1,214 @@
+//     auto && [gapped_database, gapped_query] = alignment.trace();
+//     EXPECT_EQ(std::string{gapped_database | view::to_char}, fixture.gapped_sequence1);
+//     EXPECT_EQ(std::string{gapped_query | view::to_char}, fixture.gapped_sequence2);
+// }
```
...?

h-2 replied:
But why commit dead code? If the other PRs change these lines anyway, why can't they add the code in the first place?

rrahn replied:
Well, well, I'll see that I remove it.

eseiler commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,168 @@
+     * \details
+     *
+     * The configuration is copied once to the heap during construction and maintained by a std::shared_ptr.
+     * The configuration is not passed to the function-call-operator of this function object, in order to avoid
```
```suggestion
     * The configuration is not passed to the function-call-operator of this function object in order to avoid
```

eseiler commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,168 @@
+
+        // Initialize cache variables to keep frequently used variables close to the CPU registers.
+        auto cache = this->setup_cache(get<align_cfg::gap>(*cfg_ptr).value);
+        // auto cache = this->setup_cache(gap_scheme{gap_score{-1}, gap_open_score{-10}});
```
Debug stuff?

eseiler commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,168 @@
+
+protected:
+
+    /*!\brief Initializes the first column of the dynamic programming matrix.
```
It should be BE everywhere... so `Initialises`

eseiler commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -0,0 +1,189 @@
+        // ----------------------------------------------------------------------------
+
+        if constexpr (config_t::template exists<align_cfg::aligned_ends>())
+            throw std::domain_error{"Configuring the aligned ends is yet not supported."};
```
Just for understanding:

Why not `static_assert`?
It's static constexpr, so it needs to be known at compile time, right?

In any case, would a throw in a constexpr context occur during compiling? (Never tested it 😁 )

eseiler commented on include/seqan3/alignment/pairwise/alignment_selector.hpp:
```diff
@@ -45,102 +18,79 @@
+        requires is_type_specialisation_of_v<remove_cvref_t<config_t>, configuration>
+    static constexpr auto configure(sequences_t seq_range, config_t const & cfg)
+    {
+        using first_seq_t  = std::remove_reference_t<std::tuple_element_t<0, remove_cvref_t<decltype(*seqan3::begin(seq_range))>>>;
```
Line lengths

eseiler commented on include/seqan3/alignment/pairwise/policy/affine_gap_init_policy.hpp:
```diff
@@ -0,0 +1,103 @@
+        vt_score += gap_extend;
+    }
+
+    /*!\brief Initializes the cells in the first row of the dynamic programming matrix.
```
Actually it initialises **a** cell of the first row, no?

rrahn commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,168 @@
+ * \details
+ *
+ * Computes the actual pairwise alignment given two sequences. The correct algorithm is configured while parsing
+ * the seqan3::configuration within the detail::seqan3::configurator.
```
Ah now I see what you mean 🤦‍♂️

rrahn commented:
@h-2 @eseiler I think I addressed all issues. I added documenation for the policies in the respective all header header and enhanced in general the documentation a little bit.

rrahn commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -0,0 +1,199 @@
+    std::shared_ptr<config_t> cfg_ptr{};
+};
+
+} // namespace seqan3::detail
```
Mhmm, usually the editior does this automatically.

rrahn replied:
No all files have a new line at the end. They are not displayed in the code diff

eseiler replied:
Mhm ok.
I looked at the raw file and there was none 🤔

rrahn commented on include/seqan3/alignment/pairwise/execution/alignment_executor_two_way.hpp:
```diff
@@ -27,105 +28,131 @@ namespace seqan3::detail
-    alignment_executor_two_way & operator=(alignment_executor_two_way const &) = default;
-    alignment_executor_two_way & operator=(alignment_executor_two_way &&)      = default;
-    ~alignment_executor_two_way()                                              = default;
+    alignment_executor_two_way() = delete;
```
I had this in  my snippet but then decided that it is not worth it. So I changed it but not everywhere.

rrahn commented:
The failing test was due to time limits for the coverage builds. So everything else seems to work.

h-2 commented on include/seqan3/alignment/pairwise/policy/all.hpp:
```diff
@@ -0,0 +1,132 @@
+ *
+ * ### Existing gap init policies:
+ *
+ *  - seqan3::detail::affine_gap_init_policy
```
Grwat explanation! But is this really a costumisation point, i.e. can users just add their own policy, somehow squeeze it into the config and everything will work?

Maybe you could make it a little clearer what part of this documentation to refers to users adding custom behaviour in their app and which part only addresses developers of the library (most policies seem to be detail anyway?). And maybe mark respective things with `\if DEV` to it is only visible in the developer documentation?

rrahn replied:
Oh, indeed it needs to be put into DEV. I forot this.

The intention was to provide an easier way to customise the alignment algorithms for library developers. Because most changes will involve structural changes in the algorithm and there
are depdencies between the data structures and layouts changing one thing means adapting others. Hence, the list of policies, what they are for, which interfaces they have to obey and that I need some more descriptive classes rather than plain tuples to interface with the different policies. So there are dpendencies but in general they are designed such that they can be replaced by some other implementation. That's why I would like to keep them noted as customisation points.
